core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1289,24 +1310,51 @@ public void run() {
             channelLogger.log(ChannelLogLevel.INFO, "Address resolved: {0}", servers);
             haveBackends = true;
           }
-          final Map<String, ?> serviceConfig =
+
+          // Assuming no error in config resolution for now.
+          Map<String, ?> serviceConfig =
               config.get(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG);
-          if (serviceConfig != null && !serviceConfig.equals(lastServiceConfig)) {
-            channelLogger.log(ChannelLogLevel.INFO, "Service config changed");
+          Attributes effectiveConfig = config;
+          if (!lookUpServiceConfig || serviceConfig == null) {
+            if (serviceConfig != null) { // lookUpServiceConfig is false
+              channelLogger.log(
+                  ChannelLogLevel.INFO,
+                  "Service config from name resolver discarded by channel settings");
+            }
+            if (defaultServiceConfig != null) {
+              channelLogger.log(ChannelLogLevel.INFO, "Using default service config");
+            } else if (lastServiceConfig != null && serviceConfig == null) {
+              // no default config, no current service config, last config is non-trivial
+              channelLogger.log(ChannelLogLevel.INFO, "Service config changed to null");
+            }
+
+            if (!Objects.equal(defaultServiceConfig, serviceConfig)) {
+              effectiveConfig = Attributes.newBuilder()
+                  .setAll(config)
+                  .set(GrpcAttributes.NAME_RESOLVER_SERVICE_CONFIG, defaultServiceConfig)
+                  .build();
+            }
+            lastServiceConfig = defaultServiceConfig;
+          } else {
+            if (!serviceConfig.equals(lastServiceConfig)) {
```

Comment:
Just use reference equality.

Reply:
It looks this is preserving existing behavior. It can stay as-is. (Although in the future I expect it will be changed.)